File: processor/src/main/java/io/michaelrocks/lightsaber/processor/ClassProcessor.java
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor;

import io.michaelrocks.lightsaber.processor.analysis.Analyzer;
import io.michaelrocks.lightsaber.processor.annotations.proxy.AnnotationCreator;
import io.michaelrocks.lightsaber.processor.descriptors.InjectionTargetDescriptor;
import io.michaelrocks.lightsaber.processor.descriptors.InjectorDescriptor;
import io.michaelrocks.lightsaber.processor.descriptors.ModuleDescriptor;
import io.michaelrocks.lightsaber.processor.descriptors.ProviderDescriptor;
import io.michaelrocks.lightsaber.processor.descriptors.QualifiedMethodDescriptor;
import io.michaelrocks.lightsaber.processor.descriptors.QualifiedType;
import io.michaelrocks.lightsaber.processor.descriptors.ScopeDescriptor;
import io.michaelrocks.lightsaber.processor.generation.ClassProducer;
import io.michaelrocks.lightsaber.processor.generation.InjectorFactoryClassGenerator;
import io.michaelrocks.lightsaber.processor.generation.PackageModuleClassGenerator;
import io.michaelrocks.lightsaber.processor.generation.ProcessorClassProducer;
import io.michaelrocks.lightsaber.processor.generation.ProvidersGenerator;
import io.michaelrocks.lightsaber.processor.generation.TypeAgentsGenerator;
import io.michaelrocks.lightsaber.processor.graph.CycleSearcher;
import io.michaelrocks.lightsaber.processor.graph.DependencyGraph;
import io.michaelrocks.lightsaber.processor.graph.UnresolvedDependenciesSearcher;
import io.michaelrocks.lightsaber.processor.injection.InjectionClassFileVisitor;
import io.michaelrocks.lightsaber.processor.io.ClassFileReader;
import io.michaelrocks.lightsaber.processor.io.ClassFileWriter;
import io.michaelrocks.lightsaber.processor.validation.SanityChecker;
import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.exception.ExceptionUtils;
import org.objectweb.asm.Type;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ClassProcessor {
    private final ClassFileReader classFileReader;
    private final ClassFileWriter classFileWriter;
    private final List<File> libraries;

    private final ProcessorContext processorContext = new ProcessorContext();
    private final ClassProducer classProducer;
    private final AnnotationCreator annotationCreator;

    public ClassProcessor(final ClassFileReader classFileReader, final ClassFileWriter classFileWriter,
            final List<File> libraries) {
        this.classFileWriter = classFileWriter;
        this.classFileReader = classFileReader;
        this.libraries = new ArrayList<>(libraries);

        this.classProducer = new ProcessorClassProducer(classFileWriter, processorContext);
        this.annotationCreator = new AnnotationCreator(processorContext, classProducer);
    }

    public void processClasses() throws IOException {
        performAnalysis();
        composePackageModules();
        composeInjectors();
        processorContext.dump();
        validateDependencyGraph();
        generateGlobalModule();
        generateProviders();
        generateInjectorFactory();
        generateInjectors();
        copyAndPatchClasses();
    }

    private void performAnalysis() throws IOException {
        final Analyzer analyzer = new Analyzer(processorContext);
        analyzer.analyze(classFileReader, libraries);
        new SanityChecker(processorContext).performSanityChecks();
        checkErrors();
    }

    private void composePackageModules() {
        final Map<String, ModuleDescriptor.Builder> moduleBuilders = new HashMap<>();
        for (final InjectionTargetDescriptor providableTarget : processorContext.getProvidableTargets()) {
            final Type providableTargetType = providableTarget.getTargetType();
            final QualifiedMethodDescriptor providableTargetConstructor = providableTarget.getInjectableConstructor();

            final String packageName = FilenameUtils.getPath(providableTargetType.getInternalName());
            ModuleDescriptor.Builder moduleBuilder = moduleBuilders.get(packageName);
            if (moduleBuilder == null) {
                final Type moduleType = processorContext.getPackageModuleType(packageName);
                moduleBuilder = new ModuleDescriptor.Builder(moduleType);
                moduleBuilders.put(packageName, moduleBuilder);
            }

            final Type providerType = Type.getObjectType(
                    providableTargetType.getInternalName() + "$$Provider");
            final QualifiedType providableType = new QualifiedType(providableTarget.getTargetType());
            final ScopeDescriptor scope = providableTarget.getScope();
            final Type delegatorType = scope != null ? scope.getProviderType() : null;
            final ProviderDescriptor provider =
                    new ProviderDescriptor(providerType, providableType, providableTargetConstructor,
                            moduleBuilder.getModuleType(), delegatorType);

            moduleBuilder.addProvider(provider);
        }

        for (final ModuleDescriptor.Builder moduleBuilder : moduleBuilders.values()) {
            processorContext.addPackageModule(moduleBuilder.build());
        }
    }

    private void composeInjectors() {
        for (final InjectionTargetDescriptor injectableTarget : processorContext.getInjectableTargets()) {
            final Type injectorType =
                    Type.getObjectType(injectableTarget.getTargetType().getInternalName() + "$$Agent");
            final InjectorDescriptor injector = new InjectorDescriptor(injectorType, injectableTarget);
            processorContext.addInjector(injector);
        }
    }

    private void validateDependencyGraph() throws ProcessingException {
        final DependencyGraph dependencyGraph = new DependencyGraph(processorContext, processorContext.getModules());

        final UnresolvedDependenciesSearcher unresolvedDependenciesSearcher =
                new UnresolvedDependenciesSearcher(dependencyGraph);
        final Collection<QualifiedType> unresolvedDependencies =
                unresolvedDependenciesSearcher.findUnresolvedDependencies();
        for (final QualifiedType unresolvedDependency : unresolvedDependencies) {
            processorContext.reportError(
                    new ProcessingException("Unresolved dependency: " + unresolvedDependency));
        }

        final CycleSearcher cycleSearcher = new CycleSearcher(dependencyGraph);
        final Collection<QualifiedType> cycles = cycleSearcher.findCycles();
        for (final QualifiedType cycle : cycles) {
            processorContext.reportError(
                    new ProcessingException("Cycled dependency: " + cycle));
        }

        checkErrors();
    }

    private void generateGlobalModule() throws ProcessingException {
        final PackageModuleClassGenerator packageModuleClassGenerator =
                new PackageModuleClassGenerator(classProducer, processorContext);
        for (final ModuleDescriptor packageModule : processorContext.getPackageModules()) {
            packageModuleClassGenerator.generatePackageModule(packageModule);
        }
        checkErrors();
    }

    private void generateProviders() throws ProcessingException {
        final ProvidersGenerator providersGenerator =
                new ProvidersGenerator(classProducer, processorContext, annotationCreator);
        providersGenerator.generateProviders();
        checkErrors();
    }

    private void generateInjectorFactory() throws ProcessingException {
        final InjectorFactoryClassGenerator injectorFactoryClassGenerator =
                new InjectorFactoryClassGenerator(classProducer, processorContext);
        injectorFactoryClassGenerator.generateInjectorFactory();
        checkErrors();
    }

    private void generateInjectors() throws ProcessingException {
        final TypeAgentsGenerator typeAgentsGenerator =
                new TypeAgentsGenerator(classProducer, processorContext, annotationCreator);
        typeAgentsGenerator.generateInjectors();
        checkErrors();
    }

    private void copyAndPatchClasses() throws IOException {
        final InjectionClassFileVisitor injectionVisitor =
                new InjectionClassFileVisitor(classFileWriter, processorContext, annotationCreator);
        classFileReader.accept(injectionVisitor);
    }

    private void checkErrors() throws ProcessingException {
        if (processorContext.hasErrors()) {
            throw new ProcessingException(composeErrorMessage());
        }
    }

    private String composeErrorMessage() {
        final StringBuilder builder = new StringBuilder();
        for (final Map.Entry<String, List<Exception>> entry : processorContext.getErrors().entrySet()) {
            final String path = entry.getKey();
            for (final Exception error : entry.getValue()) {
                builder
                        .append(System.lineSeparator())
                        .append(path)
                        .append(": ")
                        .append(error.getMessage())
                        .append(System.lineSeparator())
                        .append(ExceptionUtils.getStackTrace(error));
            }
        }
        return builder.toString();
    }
}

File: processor/src/main/kotlin/io/michaelrocks/lightsaber/processor/ClassProcessor.kt
/*
 * Copyright 2018 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor

import io.michaelrocks.grip.Grip
import io.michaelrocks.grip.GripFactory
import io.michaelrocks.lightsaber.processor.analysis.Analyzer
import io.michaelrocks.lightsaber.processor.commons.StandaloneClassWriter
import io.michaelrocks.lightsaber.processor.commons.closeQuietly
import io.michaelrocks.lightsaber.processor.compiler.JavaToolsCompiler
import io.michaelrocks.lightsaber.processor.generation.Generator
import io.michaelrocks.lightsaber.processor.injection.Patcher
import io.michaelrocks.lightsaber.processor.io.DirectoryFileSink
import io.michaelrocks.lightsaber.processor.io.FileSource
import io.michaelrocks.lightsaber.processor.io.IoFactory
import io.michaelrocks.lightsaber.processor.logging.getLogger
import io.michaelrocks.lightsaber.processor.model.Component
import io.michaelrocks.lightsaber.processor.model.InjectionContext
import io.michaelrocks.lightsaber.processor.model.InjectionPoint
import io.michaelrocks.lightsaber.processor.model.InjectionTarget
import io.michaelrocks.lightsaber.processor.model.ProvisionPoint
import io.michaelrocks.lightsaber.processor.validation.Validator
import org.objectweb.asm.ClassReader
import org.objectweb.asm.ClassWriter
import java.io.Closeable
import java.io.File

class ClassProcessor(
    private val inputs: List<File>,
    private val outputs: List<File>,
    private val sourcePath: File,
    private val genPath: File,
    classpath: List<File>,
    bootClasspath: List<File>
) : Closeable {
  private val logger = getLogger()

  private val grip: Grip = GripFactory.create(inputs + classpath + bootClasspath)
  private val errorReporter = ErrorReporter()

  private val fileSourcesAndSinks = inputs.zip(outputs) { input, output ->
    val source = IoFactory.createFileSource(input)
    val sink = IoFactory.createFileSink(input, output)
    source to sink
  }
  private val sourceSink = DirectoryFileSink(sourcePath)
  private val classSink = DirectoryFileSink(genPath)

  private val compiler = JavaToolsCompiler(inputs + classpath + genPath, bootClasspath, errorReporter)

  fun processClasses() {
    val context = performAnalysisAndValidation()
    context.dump()
    copyAndPatchClasses(context)
    performGeneration(context)
    performCompilation()
  }

  override fun close() {
    classSink.closeQuietly()
    sourceSink.closeQuietly()

    fileSourcesAndSinks.forEach {
      it.first.closeQuietly()
      it.second.closeQuietly()
    }
  }

  private fun performAnalysisAndValidation(): InjectionContext {
    val analyzer = Analyzer(grip, errorReporter)
    val context = analyzer.analyze(inputs)
    Validator(grip.classRegistry, errorReporter, context).validate()
    checkErrors()
    return context
  }

  private fun copyAndPatchClasses(context: InjectionContext) {
    fileSourcesAndSinks.forEach { (fileSource, fileSink) ->
      logger.debug("Copy from {} to {}", fileSource, fileSink)
      fileSource.listFiles { path, type ->
        logger.debug("Copy file {} of type {}", path, type)
        when (type) {
          FileSource.EntryType.CLASS -> {
            val classReader = ClassReader(fileSource.readFile(path))
            val classWriter = StandaloneClassWriter(
                classReader, ClassWriter.COMPUTE_MAXS or ClassWriter.COMPUTE_FRAMES, grip.classRegistry)
            val classVisitor = Patcher(classWriter, context)
            classReader.accept(classVisitor, ClassReader.SKIP_FRAMES)
            fileSink.createFile(path, classWriter.toByteArray())
          }
          FileSource.EntryType.FILE -> fileSink.createFile(path, fileSource.readFile(path))
          FileSource.EntryType.DIRECTORY -> fileSink.createDirectory(path)
        }
      }

      fileSink.flush()
    }

    checkErrors()
  }

  private fun performGeneration(context: InjectionContext) {
    val generator = Generator(grip.classRegistry, errorReporter, classSink, sourceSink)
    generator.generate(context)
    checkErrors()
  }

  private fun performCompilation() {
    compiler.compile(sourcePath, genPath)
    checkErrors()
  }

  private fun checkErrors() {
    if (errorReporter.hasErrors()) {
      throw ProcessingException(composeErrorMessage())
    }
  }

  private fun composeErrorMessage(): String {
    return errorReporter.getErrors().joinToString("\n") { it.message.orEmpty() }
  }

  private fun InjectionContext.dump() {
    components.forEach { it.dump() }
    injectableTargets.forEach { it.dump("Injectable") }
    providableTargets.forEach { it.dump("Providable") }
  }

  private fun Component.dump() {
    logger.debug("Component: {}", type)
    for (module in modules) {
      logger.debug("  Module: {}", module.type)
      for (provider in module.providers) {
        if (provider.provisionPoint is ProvisionPoint.AbstractMethod) {
          logger.debug("   Provides: {}", provider.provisionPoint.method)
        } else if (provider.provisionPoint is ProvisionPoint.Field) {
          logger.debug("   Provides: {}", provider.provisionPoint.field)
        } else {
          logger.debug("   Provides: {}", provider.provisionPoint)
        }
      }
    }

    for (subcomponent in subcomponents) {
      logger.debug("  Subcomponent: {}", subcomponent)
    }
  }

  private fun InjectionTarget.dump(name: String) {
    logger.debug("{}: {}", name, type)
    for (injectionPoint in injectionPoints) {
      when (injectionPoint) {
        is InjectionPoint.Field -> logger.debug("  Field: {}", injectionPoint.field)
        is InjectionPoint.Method -> logger.debug("  Method: {}", injectionPoint.method)
      }
    }
  }
}
